core/src/mast/node/join_node.rs

```diff
@@ -248,3 +248,57 @@ impl MastNodeExt for JoinNode {
         Self::DOMAIN
     }
 }
+
+// ------------------------------------------------------------------------------------------------
+/// Builder for creating [`JoinNode`] instances with decorators.
+pub struct JoinNodeBuilder {
+    children: [MastNodeId; 2],
+    before_enter: Vec<DecoratorId>,
+    after_exit: Vec<DecoratorId>,
+}
+
+impl JoinNodeBuilder {
+    /// Creates a new builder for a JoinNode with the specified children.
+    pub fn new(children: [MastNodeId; 2]) -> Self {
+        Self {
+            children,
+            before_enter: Vec::new(),
+            after_exit: Vec::new(),
+        }
+    }
+
+    /// Adds decorators to be executed before this node.
+    pub fn with_before_enter(mut self, decorators: impl Into<Vec<DecoratorId>>) -> Self {
```

Comment:
(posted here so that we get a thread)

> the footprint of mutablility in MastNodes shrinks: the only two places using before_enter/after_exit mutation meaningfully are serialization & assembler.

> Next, we'll better encapsulate node addition with decorators.

IIUC, subsequent PR(s) will use these builders so that ultimately, we get rid of `MastNodeExt::append_before_enter()` and `MastNodeExt::append_after_exit()`?

Reply:
Yes. The next PR, stacked on this, should make this evident.